File: sdks/deku_go_interop/interop.go
package deku_go_interop

import (
	"encoding/binary"
	"encoding/json"
	"fmt"
	"os"
	"errors"
)

var chain_to_machine *os.File
var machine_to_chain *os.File
var state map[string]interface{}

type chain_message struct {
	Type    string
	Payload interface{}
}

func check(e error) {
	if e != nil {
		panic(e)
	}
}

func log(message string) {
	// TODO: implement debug mode logging.
	if true {
		colored := fmt.Sprintf("\x1b[%dm%s\x1b[0m", 31, message)
		fmt.Fprintln(os.Stderr, colored)
	}
}

func read() []byte {
	// read the number of bytes being sent
	b := make([]byte, 8)
	_, err := chain_to_machine.Read(b)
	check(err)
	n := binary.LittleEndian.Uint16(b)
	value := make([]byte, n)
	bytes_read, err := chain_to_machine.Read(value)
	check(err)
	log(fmt.Sprintf("read %d bytes", bytes_read))
	return value
}

func write(message []byte) {
	message_length := make([]byte, 8)
	binary.LittleEndian.PutUint16(message_length, uint16(len(message)))
	machine_to_chain.Write(message_length)
	// send the message
	machine_to_chain.Write(message)
}

// Get(key) synchronously gets the bytes associated with the given
// key from the Deku state.
func Get(key string) []byte {
	value := state[key]
	if value == nil {
		return make([]byte, 0)
	}
	json_value, err := json.Marshal(value)
	check(err)
	return []byte(json_value)
}

// Set(key,value) synchronously sets a key in the Deku state
// to the given value. That value must be serializable
// with json.Marshal.
func Set(key string, value interface{}) {
	b, err := json.Marshal(value)
	check(err)
	value_str := string(b)
	message := []byte(fmt.Sprintf("[\"Set\",{\"key\":\"%s\",\"value\":%s}]", key, value_str))
	write(message);
	state[key] = value
}

type init_entry struct {
	Key  string `json:"key"`
	Value interface{} `json:"value"`
}

func init_state(initial_state map[string]interface{}) (map[string]interface{}, error) {
	message := read()
	var json_buffer []interface{}
	json.Unmarshal(message, &json_buffer)
	log(string(message))

	message_type := json_buffer[0].(string)
	
	switch message_type {
	case "Get_Initial_State":
		var initial_message []init_entry
		for key, value := range initial_state {
			initial_message = append(initial_message, init_entry{Key: key, Value: value})
		}
		b, err := json.Marshal(initial_message)
		check(err)
		init_message := fmt.Sprintf("[\"Init\", %s]", string(b))
		write([]byte(init_message))
		return init_state(initial_state);
	case "Set_Initial_State":
		vm_state := json_buffer[1].([]interface{})
		var state = make(map[string]interface{})
		for _, entry := range vm_state {
			entry := entry.([]interface{})
			key := entry[0].(string)
			value := entry[1]
			state[key] = value 
		}
		return state, nil
	default:
		return nil, errors.New("protocol not respected 4")
	} 
}

func Main(initial_state map[string]interface{}, state_transition func(sender string, tx_hash string, input []byte) (err error)) {
	log("Opening read")
	fifo_path := os.Args[1]
	log(fmt.Sprintf("fifo path: %s", fifo_path))
	machine_to_chain, _ = os.OpenFile(fifo_path+"_read", os.O_WRONLY, 0666)
	log("Opening write")
	chain_to_machine, _ = os.OpenFile(fifo_path+"_write", os.O_RDONLY, 0666)
	log("done")

	vm_state, err := init_state(initial_state)
	check(err)
	state = vm_state
	log("initialized")

	var sender_buffer []string
	var tx_hash_buffer []string
	var input_buffer []interface{}

	for {
		// TODO: replace this with a control pipe
		control := read()
		if string(control) == "\"close\"" {
			break
		}
		
		err := json.Unmarshal(read(), &sender_buffer)
		check(err)
		err = json.Unmarshal(read(), &tx_hash_buffer)
		check(err)
		err = json.Unmarshal(read(), &input_buffer)
		check(err)

		sender := sender_buffer[1]
		tx_hash := tx_hash_buffer[1]
		input, err := json.Marshal(input_buffer[1])

		log(fmt.Sprintf("Read start message: %s", string(input)))
		err = state_transition(sender, tx_hash, []byte(input))
		var end_message []byte
		if err != nil {
			end_message = []byte(fmt.Sprintf("[\"Error\", \"%s\"]", err.Error()))
		} else {
			end_message = []byte("[\"Stop\"]")
		}
		write(end_message)
	}
}

File: sdks/typescript-sdk/lib/interop.js
"use strict";
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
exports.__esModule = true;
exports.set = exports.get = exports.main = void 0;
var fs = require("fs");
var DEBUG_LOGGING = Boolean(process.env.DEKU_VM_DEBUG_LOGGING);
var log = function () {
    var message = [];
    for (var _i = 0; _i < arguments.length; _i++) {
        message[_i] = arguments[_i];
    }
    if (DEBUG_LOGGING)
        console.log.apply(console, __spreadArray(['[\x1b[32m%s\x1b[0m] %s', 'deku-vm'], message, false));
};
var machineToChain;
var chainToMachine;
var state = {}; // TODO: add a better type to JSON
/**
 * Opens two fifos, one for reading and a second one for writing
 * @returns {void}
 */
var init_fifo = function () {
    var _a;
    var fifo_path = (_a = process.argv[2]) !== null && _a !== void 0 ? _a : "/run/deku/pipe";
    log("fifo path: ".concat(fifo_path));
    log("opening read");
    machineToChain = fs.openSync("".concat(fifo_path, "_read"), "a");
    log("opening write");
    chainToMachine = fs.openSync("".concat(fifo_path, "_write"), "r");
};
/**
 * Initialize the state of the vm
 * @param initial_state the initial state provided by the vm
 * @returns the initialized state
 */
var init_state = function (initial_state) {
    var message = JSON.parse(read().toString());
    switch (message[0]) {
        case "Get_Initial_State": {
            var initial_message = Object.keys(initial_state)
                .map(function (key) { return ({ key: key, value: initial_state[key] }); });
            var init_message = "[\"Init\", ".concat(JSON.stringify(initial_message), "]");
            write(Buffer.from(init_message));
            return initial_state;
        }
        case "Set_Initial_State":
            return message[1];
        default:
            throw new Error("protocol not respected");
    }
};
/**
 * Reads the fifo and returns the result
 * @returns {Buffer} the read buffer
 */
var read = function () {
    var buffer = Buffer.alloc(8);
    fs.readSync(chainToMachine, buffer);
    var n = buffer.readUInt16LE();
    var valueBuffer = Buffer.alloc(n);
    fs.readSync(chainToMachine, valueBuffer);
    return valueBuffer;
};
// Write the given buffer to machineToChain fifo
// It will perform 2 writes, one for the size of the buffer, another one for the buffer itself
/**
 * Write the given value to the fifo
 * @param {Buffer} value the value you want to write to the fifo as a buffer
 * @returns {void} nothing
 */
var write = function (value) {
    var buffer = Buffer.alloc(8);
    buffer.writeUInt16LE(value.length);
    fs.writeFileSync(machineToChain, buffer);
    fs.writeFileSync(machineToChain, value);
};
/**
 * Set a value in the deku state for a given key
 * @param {string} key the key of the state
 * @param {string} value a string encoded in json
 */
var set = function (key, value) {
    var message = JSON.stringify(["Set", { "key": key, value: value }]);
    write(Buffer.from(message)); // TODO: check if it succeeds
    state[key] = value;
    return;
};
exports.set = set;
/**
 * Retrieves the value from the local state
 * TODO: why returning a Buffer and not a plain object ?
 * @param key the key the value
 * @returns th stored value
 */
var get = function (key) {
    var _a;
    return (_a = state[key]) !== null && _a !== void 0 ? _a : JSON.stringify(null);
};
exports.get = get;
/**
 * The main function
 * @param initial_state the initial state of your vm
 * @param state_transition the function called when there is an new input
 */
var main = function (initial_state, // TODO: add a better type for JSON values
state_transition) {
    init_fifo();
    state = init_state(initial_state);
    log("vm started");
    for (;;) {
        var raw = read().toString();
        var message = JSON.parse(raw);
        if (message === "close") {
            break;
        }
        // FIXME: this and every other log in here should be on some kind opt-in "debug logging" mode
        log("Parsed message:", message);
        var error = "";
        if (message[0] !== "Noop_transaction") {
            var transaction = message[1];
            try {
                error = state_transition(transaction);
            }
            catch (vm_err) {
                console.error(vm_err);
                error = "Unhandled exception from the VM.";
            }
        }
        else {
            log("Received noop operation");
        }
        var end_message = error ? "[\"Error\", \"".concat(error, "\"]") : '["Stop"]';
        write(Buffer.from(end_message));
    }
};
exports.main = main;
